README.md

```diff
@@ -1 +1,237 @@
+reporting is due.
+
+**usage_records:** This is a list of usage records accrued since
+the last metered bill. Each record is a dictionary containing the usage
+data, reporting_time and base_product. Where reporting_time is a timestamp
+that denotes when the usage record was reported and base_product is a CPE
+product + version string.
+
+**last_bill:** This is a dictionary containing info about the last bill.
+
+**record_id:** The ID from the CSP API for the last metered bill.
+
+**metering_time:** A timestamp denoting when the last metering occurred.
+
+**dimensions:** A dictionary containing the dimension names and units
+billed in the last metering.
+
+## Service
+
+The adapter service runs continuously based on the query interval which is
+set in the config file. For example if the query interval is set to 3600
+the adapter service will run every hour. The following actions may be taken
+when the adapter runs:
+
+### Get Usage Data
+
+Each time the adapter runs it will check for new usage data via the
+`get_usage_data` hook. This hook is will be implemented by a plugin for the
+given deployment. It is expected that the function will return a dictionary
+with usage data and it should include two keys; `base_product` and
+`reporting_time`.
+
+### Meter billing
+
+The billing interval and reporting interval determine when the adapter meters
+a bill through the CSP API. In some cases the two interval may be identical.
+In this case every metering is billing usage with the CSP. In other cases the
+reporting interval is more frequent than the billing interval. This is where
+an application is expected to provide a "heartbeat" on a shorter interval than
+the product billing period.
+
+The billing interval has two possible values; monthly and hourly. When set to
+hourly the adapter bills usage every hour. And when set to monthly the adapter
+aggregates usage based on the query interval and bills once a month.
+
+### Save data
+
+At the end of each cycle the adapter will store billing and usage data using
+the storage hookspecs. This backs up the data in case the adapter is rebooted
+or down for some amount of time.
+
+### Errors
+
+For all errors an error will be added to "errors" list in csp-config. "..." denotes an extended error message which will change based on the specific error. If the adapter is unable to access metered billing API it will also set "billing_api_access_ok" to False. If any error state is noticed in csp-config it should be communicated proactively to the customer through UI.
+
+- If the adapter cannot find the neuvector-usage CR or cannot get new records then the adapter will add an error:
+    - "Usage data retrieval failed: ..."
+- If the adapter has trouble processing the metering then the adapter will:
+    - Add error message to list. The specific message will be based on the error that occurs.
+    - Set "billing_api_access_ok" to False
+- If the adapter cannot save csp-config to configMap it will add an error:
```

Comment:
Again potentially this should be more generic, e.g. 
```suggestion
- If the adapter cannot save csp-config data (e.g. write to associated configMap in a Kubernetes deployment) it will add an error:
```